### ISSUE TITLE
Ability to zero gradients instead of reallocating for each batch

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#496]
Related to #400 

In pytorch, optimizers can call .zero_grad(), which notably just fills memory with 0s instead of deleting them.

As of now, `Gradients` is dropped once passed to optimizer, which means all gradients are deallocated.

Imagining the following loop:

```rust
let mut grads: Gradients = Default::default();
for batch in ... {
    grads.zero();
    let y = x.trace_with_grads(grads);
    let loss = ...
    grads = loss.backward()
    opt.update(&mut model, &grads);
}
```

Something would have to remove temporary tensor gradients at some point though

**Comment by coreylowman** [comment]
A giant plus over pytorch would be that Gradients stores all gradients of temporary tensors as well, meaning you wouldn't have to allocate for them anymore.

**Comment by coreylowman** [comment]
Alternatively:
```rust
let mut grads: Gradients = Default::default();
for batch in ... {
    let y = x.trace_with_grads(grads);
    let loss = ...
    grads = loss.backward()
    opt.update(&mut model, &mut grads);
}
```
where optimizer would zero gradient as its updating.

**Comment by coreylowman** [comment]
Now doing the zeroing will be easily possible with #508. Just need to figure out API for keeping grads around & destroying temporary gradients

**Comment by coreylowman** [comment]
Here are some other loops that would be used with gradient accumulation:

```rust
for i in 0.. {
    let mut grads: Gradients<E, D> = Default::default();
    for j in 0..4 {
        x.trace_with(grads);
        grads = loss.backward();
    }
    opt.update(&mut model, &grads);
}
```


```rust
let mut grads: Gradients<E, D> = Default::default();
for i in 0.. {
    grads.zero()
    for j in 0..4 {
        x.trace_with(grads);
        grads = loss.backward();
    }
    opt.update(&mut model, &grads);
}
```

```rust
let mut grads: Gradients<E, D> = Default::default();
for i in 0.. {
    x.trace_with(grads);
    grads = loss.backward();
    if i % 4 == 0 {
        opt.update(&mut model, &grads);
        grads.zero()
    }
}
```

and we could also support the normal flow, but now optimizer would take a reference to grads

**Comment by coreylowman** [comment]
It would also be nice if grads.zero() did something to free the temporary gradients, but unclear how it would do that.

We could do this with module walkers:
```rust
model.zero_grads(&mut grads);
```
and then the walker would keep track of things that hadn't been updated and free them
